icicle/polynomials/cuda_backend/kernels.cuh

```diff
@@ -0,0 +1,118 @@
+#pragma once
+#include "stdint.h"
+
+namespace polynomials {
+
+  /*============================== add/sub ==============================*/
+  template <typename T>
+  __global__ void AddSubKernel(T* a_vec, T* b_vec, int a_len, int b_len, bool add1_sub0, T* result)
+  {
+    const int tid = blockIdx.x * blockDim.x + threadIdx.x;
+    if (tid >= max(a_len, b_len)) return;
+
+    T a = tid >= a_len ? T::zero() : a_vec[tid];
+    T b = tid >= b_len ? T::zero() : b_vec[tid];
+    result[tid] = add1_sub0 ? a + b : a - b;
+  }
+
+  template <typename T>
+  __global__ void AddSingleElementInplace(T* self, T v)
+  {
+    const int tid = blockIdx.x * blockDim.x + threadIdx.x;
+    if (tid > 0) return;
+    *self = *self + v;
+  }
+
+  /*============================== degree ==============================*/
+  template <typename T>
+  __global__ void HighestNonZeroIdx(T* vec, int len, int64_t* idx)
+  {
+    *idx = -1; // zero polynomial is defined with degree -1
+    for (int64_t i = len - 1; i >= 0; --i) {
+      if (vec[i] != T::zero()) {
+        *idx = i;
+        return;
+      }
+    }
+  }
+
+  /*============================== evaluate ==============================*/
+  template <typename T>
+  __device__ T pow(T base, int exp)
```

Comment:
Yes maybe. Note that this is not compute very efficiently. The reduction is serially and so maybe Horner's rule is better since my solution requires extra memory. But again, I did not really put much effort in making it optimal at this point.